File: AndroidApp/android/app/src/main/java/com/wyre/app/TransferCrypto.kt
package com.wyre.app

import android.util.Base64
import java.io.InputStream
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.spec.ECGenParameterSpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * TransferCrypto.kt
 * ECDH P-256 key exchange + AES-256-GCM chunk encryption/decryption.
 * Uses Android's javax.crypto / java.security only — no third-party dependencies.
 *
 * Wire format per chunk:
 *   [4 bytes: payload length uint32 big-endian]
 *   [12 bytes: IV]
 *   [N bytes: AES-256-GCM ciphertext]
 *   [16 bytes: GCM authentication tag]
 */
object TransferCrypto {

    private const val IV_LENGTH = 12
    private const val TAG_LENGTH_BITS = 128
    private const val TAG_LENGTH_BYTES = 16
    private const val KEY_LENGTH_BYTES = 32

    // ── Key generation ────────────────────────────────────────────────────────

    /**
     * Generate an ephemeral ECDH P-256 key pair.
     * Returns (publicKeyDerBytes, privateKey).
     * The public key is DER-encoded SubjectPublicKeyInfo (SPKI) — safe to send over wire.
     */
    fun generateKeyPair(): Pair<ByteArray, PrivateKey> {
        val gen = KeyPairGenerator.getInstance("EC")
        gen.initialize(ECGenParameterSpec("secp256r1"), SecureRandom())
        val kp = gen.generateKeyPair()
        // getEncoded() on an EC public key returns DER SPKI format
        return Pair(kp.public.encoded, kp.private)
    }

    // ── Key derivation ────────────────────────────────────────────────────────

    /**
     * Derive a 256-bit AES-GCM key via HKDF-SHA-256.
     *
     * Both sides must call this with the same inputs (order: sender then receiver)
     * to arrive at the same key.
     *
     * Steps:
     *  1. ECDH key agreement → shared secret
     *  2. HKDF-Extract: PRK = HMAC-SHA256(salt, sharedSecret)
     *     where salt = 32 zero bytes || senderPublicKeyDer || receiverPublicKeyDer
     *  3. HKDF-Expand: OKM = HMAC-SHA256(PRK, info || 0x01)
     *     where info = "wyre-transfer-v1"
     *  4. Return first 32 bytes as the AES key
     */
    fun deriveKey(
        privateKey: PrivateKey,
        peerPublicKeyDer: ByteArray,
        senderPublicKeyDer: ByteArray,
        receiverPublicKeyDer: ByteArray,
    ): ByteArray {
        // 1. ECDH key agreement
        val peerPublicKey = decodeSpkiPublicKey(peerPublicKeyDer)
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(privateKey)
        ka.doPhase(peerPublicKey, true)
        val sharedSecret = ka.generateSecret()

        // 2. HKDF-Extract
        // salt = 32 zero bytes || senderPublicKeyDer || receiverPublicKeyDer
        val zeroSalt = ByteArray(KEY_LENGTH_BYTES) { 0 }
        val combinedSalt = zeroSalt + senderPublicKeyDer + receiverPublicKeyDer
        val prk = hmacSha256(combinedSalt, sharedSecret)

        // 3. HKDF-Expand (single block, T(1))
        val info = "wyre-transfer-v1".toByteArray(Charsets.UTF_8)
        val t1Input = info + byteArrayOf(0x01)
        val okm = hmacSha256(prk, t1Input)

        // 4. Return first 32 bytes
        return okm.copyOf(KEY_LENGTH_BYTES)
    }

    // ── Encryption ────────────────────────────────────────────────────────────

    /**
     * Encrypt one chunk.
     * Returns Triple(iv, ciphertext, tag).
     *
     * Android's AES/GCM/NoPadding Cipher.doFinal() returns ciphertext || tag
     * concatenated — we split by taking the last 16 bytes as the tag.
     */
    fun encryptChunk(key: ByteArray, plaintext: ByteArray): Triple<ByteArray, ByteArray, ByteArray> {
        val iv = ByteArray(IV_LENGTH).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(key, "AES")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, iv)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)
        val output = cipher.doFinal(plaintext)

        // output = ciphertext || tag (last TAG_LENGTH_BYTES bytes)
        val ciphertext = output.copyOf(output.size - TAG_LENGTH_BYTES)
        val tag = output.copyOfRange(output.size - TAG_LENGTH_BYTES, output.size)
        return Triple(iv, ciphertext, tag)
    }

    // ── Decryption ────────────────────────────────────────────────────────────

    /**
     * Decrypt one chunk. Throws AEADBadTagException on authentication failure.
     *
     * Android GCM expects ciphertext || tag concatenated before doFinal.
     */
    fun decryptChunk(key: ByteArray, iv: ByteArray, ciphertext: ByteArray, tag: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(key, "AES")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, iv)
        cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)
        // Concatenate ciphertext + tag — Android GCM expects them together
        val combined = ciphertext + tag
        return cipher.doFinal(combined)
    }

    // ── Wire format ───────────────────────────────────────────────────────────

    /**
     * Encode to wire format:
     * [4-byte payload length][12-byte IV][ciphertext][16-byte tag]
     */
    fun encodeChunk(iv: ByteArray, ciphertext: ByteArray, tag: ByteArray): ByteArray {
        val payloadLength = iv.size + ciphertext.size + tag.size
        val result = ByteArray(4 + payloadLength)
        // Write big-endian uint32 length
        result[0] = (payloadLength shr 24).toByte()
        result[1] = (payloadLength shr 16).toByte()
        result[2] = (payloadLength shr 8).toByte()
        result[3] = payloadLength.toByte()
        var pos = 4
        iv.copyInto(result, pos); pos += iv.size
        ciphertext.copyInto(result, pos); pos += ciphertext.size
        tag.copyInto(result, pos)
        return result
    }

    /**
     * Decode one encrypted chunk from an InputStream.
     * Returns Triple(iv, ciphertext, tag) or null on EOF.
     * Throws on malformed data or stream errors.
     */
    fun decodeChunkFromStream(inputStream: InputStream): Triple<ByteArray, ByteArray, ByteArray>? {
        // Read 4-byte length
        val lenBuf = ByteArray(4)
        val lenRead = inputStream.readNBytes(lenBuf, 0, 4)
        if (lenRead == 0) return null // clean EOF
        if (lenRead < 4) throw java.io.IOException("Unexpected EOF reading chunk length")

        val payloadLength = ((lenBuf[0].toInt() and 0xFF) shl 24) or
                            ((lenBuf[1].toInt() and 0xFF) shl 16) or
                            ((lenBuf[2].toInt() and 0xFF) shl 8) or
                             (lenBuf[3].toInt() and 0xFF)

        if (payloadLength < IV_LENGTH + TAG_LENGTH_BYTES) {
            throw java.io.IOException("Chunk payload too small: $payloadLength bytes")
        }

        val payload = ByteArray(payloadLength)
        val payloadRead = inputStream.readNBytes(payload, 0, payloadLength)
        if (payloadRead < payloadLength) throw java.io.IOException("Unexpected EOF reading chunk payload")

        val iv = payload.copyOf(IV_LENGTH)
        val tag = payload.copyOfRange(payloadLength - TAG_LENGTH_BYTES, payloadLength)
        val ciphertext = payload.copyOfRange(IV_LENGTH, payloadLength - TAG_LENGTH_BYTES)

        return Triple(iv, ciphertext, tag)
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun decodeSpkiPublicKey(der: ByteArray): PublicKey {
        val spec = X509EncodedKeySpec(der)
        return KeyFactory.getInstance("EC").generatePublic(spec)
    }
}

File: DesktopApp/src/main/crypto/TransferCrypto.ts
/**
 * TransferCrypto.ts
 * ECDH P-256 key exchange + AES-256-GCM chunk encryption/decryption.
 * Uses Node.js built-in 'crypto' module only — no third-party dependencies.
 */

import {
  generateKeyPairSync,
  createECDH,
  createCipheriv,
  createDecipheriv,
  randomBytes,
  hkdfSync,
} from 'crypto';

export interface KeyPair {
  publicKeyDer: Buffer;  // DER SubjectPublicKeyInfo — safe to send over wire
  privateKey: Buffer;    // raw private key bytes — never leave this process
}

export interface EncryptedChunk {
  iv: Buffer;         // 12 bytes
  ciphertext: Buffer; // encrypted data
  tag: Buffer;        // 16 bytes GCM auth tag
}

export class TransferCrypto {
  /** Generate an ephemeral ECDH P-256 key pair for one transfer. */
  static generateKeyPair(): KeyPair {
    const { publicKey, privateKey } = generateKeyPairSync('ec', {
      namedCurve: 'prime256v1',
      publicKeyEncoding:  { type: 'spki',  format: 'der' },
      privateKeyEncoding: { type: 'sec1',  format: 'der' },
    });

    return {
      publicKeyDer: publicKey as unknown as Buffer,
      privateKey:   privateKey as unknown as Buffer,
    };
  }

  /**
   * Derive a 256-bit AES-GCM key from two ECDH public keys using HKDF-SHA-256.
   * Both sides must call this with the same inputs (order: sender then receiver)
   * to arrive at the same key.
   */
  static deriveKey(
    ownPrivateKey: Buffer,
    peerPublicKeyDer: Buffer,
    senderPublicKeyDer: Buffer,
    receiverPublicKeyDer: Buffer,
  ): Buffer {
    // Perform ECDH key agreement
    const ecdh = createECDH('prime256v1');

    // Import our private key from DER (sec1 format)
    // Node's createECDH accepts the raw private key scalar as a Buffer
    // We need to extract the raw private key from the SEC1 DER encoding.
    // SEC1 DER structure: SEQUENCE { INTEGER version, OCTET STRING privateKey, ... }
    // The private key bytes start at offset 7 in a standard P-256 SEC1 DER.
    const rawPrivKey = extractRawPrivateKeyFromSec1Der(ownPrivateKey);
    ecdh.setPrivateKey(rawPrivKey);

    // Import peer's public key from DER (SPKI format)
    // Extract the raw uncompressed public key point (65 bytes) from SPKI DER.
    const rawPeerPubKey = extractRawPublicKeyFromSpkiDer(peerPublicKeyDer);
    const sharedSecret = ecdh.computeSecret(rawPeerPubKey);

    // HKDF-SHA-256: salt = 32 zero bytes, info = "wyre-transfer-v1"
    const salt = Buffer.alloc(32, 0);
    const info = Buffer.from('wyre-transfer-v1');

    // Deterministic salt derived from both public keys so both sides agree
    // (the salt is the concatenation of sender + receiver public keys hashed)
    const combinedSalt = Buffer.concat([senderPublicKeyDer, receiverPublicKeyDer]);
    const hkdfSalt = Buffer.concat([salt, combinedSalt]);

    const derived = hkdfSync('sha256', sharedSecret, hkdfSalt, info, 32);
    return Buffer.from(derived);
  }

  /** Encrypt one chunk. Returns IV + ciphertext + GCM tag as separate Buffers. */
  static encryptChunk(key: Buffer, plaintext: Buffer): EncryptedChunk {
    const iv = randomBytes(12);
    const cipher = createCipheriv('aes-256-gcm', key, iv);
    const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
    const tag = cipher.getAuthTag();
    return { iv, ciphertext, tag };
  }

  /** Decrypt one chunk. Verifies GCM tag — throws if authentication fails. */
  static decryptChunk(key: Buffer, iv: Buffer, ciphertext: Buffer, tag: Buffer): Buffer {
    const decipher = createDecipheriv('aes-256-gcm', key, iv);
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  }

  /**
   * Encode an encrypted chunk to wire format:
   * [4-byte length][12-byte IV][ciphertext][16-byte tag]
   */
  static encodeChunk(chunk: EncryptedChunk): Buffer {
    const payloadLength = chunk.iv.length + chunk.ciphertext.length + chunk.tag.length;
    const lengthBuf = Buffer.allocUnsafe(4);
    lengthBuf.writeUInt32BE(payloadLength, 0);
    return Buffer.concat([lengthBuf, chunk.iv, chunk.ciphertext, chunk.tag]);
  }

  /**
   * Read one encrypted chunk from a Buffer at a given offset.
   * Returns { iv, ciphertext, tag, bytesConsumed }.
   */
  static decodeChunk(
    buf: Buffer,
    offset: number,
  ): { iv: Buffer; ciphertext: Buffer; tag: Buffer; bytesConsumed: number } {
    if (buf.length - offset < 4) {
      throw new Error('Buffer too small to read chunk length');
    }
    const payloadLength = buf.readUInt32BE(offset);
    const totalNeeded = 4 + payloadLength;
    if (buf.length - offset < totalNeeded) {
      throw new Error('Buffer too small to read full chunk');
    }

    const ivStart = offset + 4;
    const iv = buf.slice(ivStart, ivStart + 12);
    const tagStart = offset + 4 + payloadLength - 16;
    const tag = buf.slice(tagStart, tagStart + 16);
    const ciphertext = buf.slice(ivStart + 12, tagStart);

    return { iv, ciphertext, tag, bytesConsumed: totalNeeded };
  }
}

// ─── DER parsing helpers ──────────────────────────────────────────────────────

/**
 * Extract the raw 32-byte private key scalar from a SEC1 DER-encoded P-256 key.
 *
 * SEC1 DER structure for P-256:
 *   30 77                    SEQUENCE
 *     02 01 01               INTEGER version = 1
 *     04 20 <32 bytes>       OCTET STRING (private key)
 *     ...
 *
 * The private key octet string starts after: 30 xx 02 01 01 04 20
 * That is: tag(1) + len(1) + tag(1) + len(1) + val(1) + tag(1) + len(1) = 7 bytes
 */
function extractRawPrivateKeyFromSec1Der(der: Buffer): Buffer {
  // Walk the DER to find the OCTET STRING containing the private key
  // SEQUENCE tag = 0x30, INTEGER tag = 0x02, OCTET STRING tag = 0x04
  let pos = 0;

  // Skip outer SEQUENCE
  if (der[pos] !== 0x30) throw new Error('Expected SEQUENCE in SEC1 DER');
  pos++;
  pos += derLengthBytes(der, pos); // skip length field

  // Skip INTEGER (version = 1)
  if (der[pos] !== 0x02) throw new Error('Expected INTEGER in SEC1 DER');
  pos++;
  const intLen = derReadLength(der, pos);
  pos += derLengthBytes(der, pos);
  pos += intLen;

  // Read OCTET STRING (private key)
  if (der[pos] !== 0x04) throw new Error('Expected OCTET STRING in SEC1 DER');
  pos++;
  const keyLen = derReadLength(der, pos);
  pos += derLengthBytes(der, pos);

  return der.slice(pos, pos + keyLen);
}

/**
 * Extract the raw 65-byte uncompressed public key point from a SPKI DER-encoded P-256 key.
 *
 * SPKI DER structure:
 *   30 59                    SEQUENCE
 *     30 13                  SEQUENCE (AlgorithmIdentifier)
 *       06 07 ...            OID (ecPublicKey)
 *       06 08 ...            OID (prime256v1)
 *     03 42 00 <65 bytes>    BIT STRING (public key point, uncompressed)
 */
function extractRawPublicKeyFromSpkiDer(der: Buffer): Buffer {
  let pos = 0;

  // Skip outer SEQUENCE
  if (der[pos] !== 0x30) throw new Error('Expected SEQUENCE in SPKI DER');
  pos++;
  pos += derLengthBytes(der, pos);

  // Skip AlgorithmIdentifier SEQUENCE
  if (der[pos] !== 0x30) throw new Error('Expected AlgorithmIdentifier SEQUENCE in SPKI DER');
  pos++;
  const algLen = derReadLength(der, pos);
  pos += derLengthBytes(der, pos);
  pos += algLen;

  // Read BIT STRING
  if (der[pos] !== 0x03) throw new Error('Expected BIT STRING in SPKI DER');
  pos++;
  const bitLen = derReadLength(der, pos);
  pos += derLengthBytes(der, pos);

  // BIT STRING has a leading "unused bits" byte (always 0x00 for EC keys)
  pos++; // skip unused-bits byte
  const keyLen = bitLen - 1;

  return der.slice(pos, pos + keyLen);
}

/** Read a DER length value at the given position. */
function derReadLength(buf: Buffer, pos: number): number {
  const first = buf[pos]!;
  if (first < 0x80) return first;
  const numBytes = first & 0x7f;
  let len = 0;
  for (let i = 0; i < numBytes; i++) {
    len = (len << 8) | buf[pos + 1 + i]!;
  }
  return len;
}

/** Return the number of bytes the DER length field occupies at the given position. */
function derLengthBytes(buf: Buffer, pos: number): number {
  const first = buf[pos]!;
  if (first < 0x80) return 1;
  return 1 + (first & 0x7f);
}
